base/compiler/typelimits.jl

```diff
@@ -559,17 +559,24 @@ function tmeet(@nospecialize(v), @nospecialize(t))
             return Bottom
         end
         @assert widev <: Tuple
-        new_fields = Vector{Any}(undef, length(v.fields))
-        for i = 1:length(new_fields)
-            if isa(v.fields[i], Core.TypeofVararg)
-                new_fields[i] = v.fields[i]
-            else
-                new_fields[i] = tmeet(v.fields[i], widenconst(getfield_tfunc(t, Const(i))))
-                if new_fields[i] === Bottom
-                    return Bottom
-                end
+        if isa(ti, DataType) && ti.name === Tuple.name
+            num_fields = length(ti.parameters)
+            isva = isvatuple(ti)
+        else
+            nfields_ti = nfields_tfunc(ti)
+            isva = !isa(nfields_ti, Const)
+            num_fields = isva ? length(v.fields) : (nfields_ti::Const).val
+        end
```

Comment:
I'm not too happy about this. What I really want to know is: What is the minimum number of fields of `ti` and can there be more (vararg)? I get that for the simple case (`if` branch), but not quite for the `else` branch. This is likely not the only place we need this, so I probably just missed an elegant solution to this?

Reply:
IIUC, this means that if we started with `Tuple{Int, Vararg{Float64}}`, and ended up with something shorter somehow, that we'd compute `Tuple{Vararg{Int}}` below, instead of `Tuple{Vararg{Int, Float64}}`. So perhaps we should write this with `max(length(v.fields), (nfields_ti::Const).val::Int)`?

Reply:
Let me illustrate with an example: Let
```julia
v == PartialStruct(Tuple{Int64, Vararg{Number}}, Any[Core.Const(1), Vararg{Number}]
```
and
```julia
t == Union{Tuple{Int,Int,Int},Tuple{Int,Float64,Float64,Float64}}
```
then we get `ti == t`. And now I'd like to figure out that `ti` has at least 3 fields. At present `tmeet` (with this PR) gives
```julia
Core.PartialStruct(Tuple{Int64, Vararg{Union{Float64, Int64}}}, Any[Core.Const(1), Vararg{Union{Float64, Int64}}])
```
but it could even be
```julia
Core.PartialStruct(Tuple{Int64, Union{Float64, Int64}, Union{Float64, Int64}, Vararg{Float64}}, Any[Core.Const(1), Union{Float64, Int64}, Union{Float64, Int64}, Vararg{Union{Float64, Int64}}])
```
(But maybe that's a bit too greedy and we leave it as is.)

Reply:
`datatype_min_ninitialized` now gives you the "minimum number of fields" query, and `isknownlength` gives you the `isva` query (might want to add UnionAll and Union handling though, since currently they only accept DataType)

Reply:
Do you think it is worthwhile to extend these for `UnionAll` and `Union` (and use them here)? Maybe postpone to a follow-up PR?